PurchasesCoreSwift/Caching/DeviceCache.swift

```diff
@@ -0,0 +1,491 @@
+        writeCache {
+            self.threadUnsafeClearPurchaserInfoCacheTimestamp(appUserID: appUserID)
+            self.userDefaults.removeObject(forKey: CacheKeyBases.purchaserInfoAppUserDefaults + appUserID)
+        }
+    }
+
+    @objc(setPurchaserInfoCacheTimestampToNowForAppUserID:)
+    open func setPurchaserInfoCacheTimestampToNow(appUserID: String) {
+        writeCache {
+            self.threadUnsafeSetPurchaserInfoCacheTimestampToNow(appUserID: appUserID)
+        }
+    }
+
+    // MARK: - offerings
+    @objc(cacheOfferings:) open func cache(offerings: Offerings) {
+        offeringsCachedObject.cache(instance: offerings)
+    }
+
+    @objc open func isOfferingsCacheStale(isAppBackgrounded: Bool) -> Bool {
+        let cacheDurationInSeconds = cacheDurationInSeconds(isAppBackgrounded: isAppBackgrounded)
+        return offeringsCachedObject.isCacheStale(durationInSeconds: cacheDurationInSeconds)
+    }
+
+    @objc open func clearOfferingsCacheTimestamp() {
+        offeringsCachedObject.clearCacheTimestamp()
+    }
+
+    @objc open func setOfferingsCacheTimestampToNow() {
+        offeringsCachedObject.updateCacheTimestamp(date: Date())
+    }
+
+    // MARK: - subscriber attributes
+
+    @objc(storeSubscriberAttribute:appUserID:)
+    open func store(subscriberAttribute: SubscriberAttribute, appUserID: String) {
+        store(subscriberAttributesByKey: [subscriberAttribute.key: subscriberAttribute], appUserID: appUserID)
+    }
+
+    @objc(storeSubscriberAttributes:appUserID:)
+    open func store(subscriberAttributesByKey: [String: SubscriberAttribute], appUserID: String) {
+        guard !subscriberAttributesByKey.isEmpty else {
+            return
+        }
+
+        writeCache {
+            var groupedSubscriberAttributes = self.threadUnsafeStoredAttributesForAllUsers
+            var subscriberAttributesForAppUserID = groupedSubscriberAttributes[appUserID] as? [String: Any] ?? [:]
+            for (key, attributes) in subscriberAttributesByKey {
+                subscriberAttributesForAppUserID[key] = attributes.asDictionary()
+            }
+            groupedSubscriberAttributes[appUserID] = subscriberAttributesForAppUserID
+            self.userDefaults.setValue(groupedSubscriberAttributes, forKey: .subscriberAttributes)
+        }
+    }
+
+    @objc(subscriberAttributeWithKey:appUserID:)
+    open func subscriberAttribute(attributeKey: String, appUserID: String) -> SubscriberAttribute? {
+        return readCache { self.threadUnsafeStoredSubscriberAttributes(appUserID: appUserID)[attributeKey] }
+    }
+
+    // Threadsafe using accessQueue, however accessQueue is not reentrant. If you're calling this from somewhere
+    // that ends up in the accessQueue, it will deadlock. Instead, you'll want to call the version that isn't wrapped
+    // in a readCache{} block: threadUnsafeUnsyncedAttributesByKey
+    @objc open func unsyncedAttributesByKey(appUserID: String) -> [String: SubscriberAttribute] {
```

Comment:
I feel like the best way to go around these issues will be to refactor this into async calls and just not assume that the results will come in right away. 

Obviously that's a bigger refactor, but it's the approach that swift actors take and I think it'll simplify a lot of what we're doing.

Reply:
Yeah, I was thinking exactly the same thing! Especially since the new `async functions` feature came out, it should simplify a lot of our logic too!